### PR TITLE
Maintenance

### PR DESCRIPTION
**Pull request by LilianBoulard**
This PR aims at improving the overall quality of the code and doc.

It has several purposes:
- Correct typos
- Reword unclear sentences
- Minor updates to the doc
- Some minor structural improvements, such as moving some functions to suiting modules
- Rename some variables for better readability
- Use modern language features for better readability and performance
- Simplify the code, while leaving functionalities intact (no bug-fixes)
- Make extensive use of type hinting, which serves two purposes:
  - Make the code easier to work with, especially when working with IDEs that support type hinting
  - Make the functions more efficient and less error-prone when using tools that enforce types, such as MyPy

In general, these are rather small modifications for which making unique PRs would be kind of overkill.

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 1]
```diff
@@ -1,19 +1,19 @@
 ---
 name: Feature request
-about: Got an idea for a new feature, or changing an existing one? This is the place.
+about: Got an idea for a new feature, or changing an existing one ? This is the place.
```
Comment (LeoGrin) [inline .github/ISSUE_TEMPLATE/feature_request.md:3]: Not important, but I think the previous one was correct

File: dirty_cat/target_encoder.py [hunk 2]
```diff
@@ -239,9 +260,10 @@ def transform(self, X):
             if not np.all(valid_mask):
                 if self.handle_unknown == 'error':
                     diff = np.unique(X[~valid_mask, i])
-                    msg = ("Found unknown categories {0} in column {1}"
-                           " during transform".format(diff, i))
-                    raise ValueError(msg)
+                    raise ValueError(
+                        "Found unknown categories {diff} in column {i} "
```
Comment (LeoGrin) [inline dirty_cat/target_encoder.py:264]: Shouldn't it be a f string?

File: examples/06_dimension_reduction_and_performance.py [hunk 3]
```diff
@@ -1,46 +1,59 @@
...
-with online estimation. As a result is it more scalable than the
-SimilarityEncoder, and should be preferred in large-scale settings.
+
+.. note::
+    The |Gap| naturally does data reduction and comes with online estimation.
+    As a result, is it more scalable than the |SE|,
+    and should be preferred in large-scale settings.
+
+
+.. |SE| replace:: :class:`~dirty_cat.SimilarityEncoder`
```
Comment (LeoGrin) [inline examples/06_dimension_reduction_and_performance.py:16]: Cool stuff :)

**Review by LeoGrin — COMMENTED** [review]
Seems very useful thanks!

File: dirty_cat/target_encoder.py (hunk 2 above)
Comment (LilianBoulard) [reply dirty_cat/target_encoder.py:264]: Nice catch !

**Comment by LilianBoulard** [comment]
Okay, so the tests pass, but [the build for the minimal requirements](https://github.com/dirty-cat/dirty_cat/runs/7325202235?check_suite_focus=true) takes waaaay too long to complete (more than 15 minutes!).

File: examples/06_dimension_reduction_and_performance.py [hunk 4]
```diff
@@ -50,29 +63,32 @@ def wrapped_func(*args, **kwargs):
...
 
+# Remove the missing lines in X
+na_mask: pd.DataFrame = X.isna()
+X = X.dropna(axis=0).reset_index()
+
 y = open_payments.y
-# Combine boolean masks
+# Combine boolean masks ; TODO: simplify
 na_mask = reduce(lambda acc, col: acc | na_mask[col],
                  na_mask.columns, na_mask[na_mask.columns[0]])
```
[the remark this answers is not in the document]
Comment (jovan-stojanovic) [reply examples/06_dimension_reduction_and_performance.py:86]: na_mask = na_mask.any(axis=1) works fine I think

**Comment by LilianBoulard** [comment]
It turns out the reason the CI takes so long is because of the minimal version of SciPy. Curently, it tries to build it from scratch.

In order to fix that, we should upgrade to [SciPy 1.4.0](https://github.com/scipy/scipy/releases/tag/v1.4.0) (released December 2019) and to [NumPy 1.17.3](https://github.com/numpy/numpy/releases/tag/v1.17.3) (released October 2019) the minimum requirement for the SciPy version.
What do you think ?

I'll push a commit applying this suggestion (see if it works); we can still rollback if necessary :)

**Comment by GaelVaroquaux** [comment]
> In order to fix that, we should upgrade to SciPy 1.4.0 (released December 2019) and to NumPy 1.17.3 (released October 2019) the minimum requirement for the SciPy version.
> What do you think ?

Check that these version or version more recent ones are available in the latest LTS of Ubuntu. You can do this using: https://packages.ubuntu.com/

**Comment by LilianBoulard** [comment]
Yep, scipy 1.8.0 and numpy 1.21.5 (listed [here](https://packages.ubuntu.com/jammy/python/) under `python3-scipy` and `python3-numpy`).

**Comment by GaelVaroquaux** [comment]
> Yep, scipy 1.8.0 and numpy 1.21.5 (listed here under python3-scipy and python3-numpy).

Go for it, then!

**Comment by LilianBoulard** [comment]
We are having an issue with the CI. I think it is linked to https://github.com/sphinx-gallery/sphinx-gallery/issues/967, I left a message there. I'll investigate further when I have a little more time !
I also want to make a last pass on the code before marking this PR as reviewable.

**Comment by LilianBoulard** [comment]
Okay, this was the pass I mentioned I'd do. I think this is pretty much done, there are just a couple of things to fix (I'm especially thinking of the TODO to remove in example 6).
CircleCI still fails, and I'm not too sure why yet.

**Comment by LilianBoulard** [comment]
The CI is clear, and the PR is ready for review !

**Comment by GaelVaroquaux** [comment]
Conflicts need to be fixed.

Also, what was the motivation for the directory name "cimg"? I have not seen it used in other projects (maybe it's because I have not followed well).

File: dirty_cat/datetime_encoder.py [hunk 5]
```diff
@@ -2,50 +2,70 @@
...
+    to the provided timezone.
 
     Parameters
     ----------
-    extract_until : {"year", "month", "day", "hour", "minute", "second", "millisecond", "microsecond", "nanosecond"}, default="hour"
-        Extract up to this granularity. If all features have not been extracted, add the "total_time" feature, which contains
-        the time to epoch (in seconds).
-        For instance, if you specify "day", only "year", "month", "day" and "total_time" features will be created.
+    extract_until : {"year", "month", "day", "hour", "minute", "second",
+                    "millisecond", "microsecond", "nanosecond"}, default="hour"
```
Comment (GaelVaroquaux) [inline dirty_cat/datetime_encoder.py:37]: Did you check that this renders well on the generated doc?

File: dirty_cat/gap_encoder.py [hunk 6]
```diff
@@ -50,12 +57,28 @@ class GapEncoderColumn(BaseEstimator, TransformerMixin):
...
-                 rescale_W=True, max_iter_e_step=20):
+    rho_: float
+    H_dict_: Dict[np.array, np.array]
+
+    def __init__(self,
+                 n_components: int = 10,
+                 batch_size: int = 128,
+                 gamma_shape_prior: float = 1.1,
+                 gamma_scale_prior: float = 1.0,
+                 rho: float = .95,
```
Comment (GaelVaroquaux) [inline dirty_cat/gap_encoder.py:68]: It used to be (in PEP8) that for keyword arguments, there should not be a space before and after the equal sign.

Can you check that the present formatting is valid with style guidelines?

Comment (LilianBoulard) [reply dirty_cat/gap_encoder.py:68]: Yes, this is valid :)

File: dirty_cat/datetime_encoder.py [hunk 7]
```diff
@@ -86,76 +108,93 @@ def _extract_from_date(self, date_series, feature):
...
-        self._to_extract = TIME_LEVELS[:TIME_LEVELS.index(self.extract_until) + 1]
+        # Columns to extract for each column,
+        # before taking into account constant columns
+        self._to_extract = TIME_LEVELS[:TIME_LEVELS.index(
+            self.extract_until) + 1]
         if self.add_day_of_the_week:
             self._to_extract.append("dayofweek")
         if isinstance(X, pd.DataFrame):
-            self.col_names_ = X.columns
+            self.col_names_ = X.columns.to_list()
```
[the remark this answers is not in the document]
Comment (LilianBoulard) [reply dirty_cat/datetime_encoder.py:143]: Mainly to fit the docstring description of the variable (`Optional[List[str]]`), `X.columns` is a `pandas.Index`.

File: dirty_cat/datetime_encoder.py (hunk 5 above)
Comment (LilianBoulard) [reply dirty_cat/datetime_encoder.py:37]: Good catch, I've also noticed class variables are added in the generated doc, which is ugly.

File: dirty_cat/minhash_encoder.py [hunk 8]
```diff
@@ -47,90 +51,86 @@ class MinHashEncoder(BaseEstimator, TransformerMixin):
...
+
         Parameters
         ----------
         string : str
             The string to encode.
+
         Returns
         -------
-        array, shape (n_components, )
+        np.array of shape (n_components, )
```
Comment (GaelVaroquaux) [inline dirty_cat/minhash_encoder.py:133]: Is this consistent with numpy doc standard?

Long term, we should add a check of consistency, as scikit-learn has recently done.

File: dirty_cat/similarity_encoder.py [hunk 9]
```diff
@@ -30,58 +34,68 @@
...
+        of size len(X) x len(cats)
 
     ngram_sim(s_i, s_j) =
         ||min(ci, cj)||_1 / (||ci||_1 + ||cj||_1 - ||min(ci, cj)||_1)
     """
     min_n, max_n = ngram_range
     unq_X = np.unique(X)
-    cats = np.array([' %s ' % cat for cat in cats])
-    unq_X_ = np.array([' %s ' % x for x in unq_X])
+    cats = np.array([f' {cat} ' for cat in cats])
```
Comment (GaelVaroquaux) [inline dirty_cat/similarity_encoder.py:97]: Did you time the difference (with large arrays)? I worry that one might be faster than the other.

File: dirty_cat/minhash_encoder.py (hunk 8 above)
Comment (LilianBoulard) [reply dirty_cat/minhash_encoder.py:133]: Indeed ! I'll leave it like that for now, but you're right that we need a consistency check. I've opened #305 to address that later on.

File: dirty_cat/similarity_encoder.py (hunk 9 above)
Comment (LilianBoulard) [reply dirty_cat/similarity_encoder.py:97]: Indeed, there is a difference as highlighted [here](https://colab.research.google.com/drive/1rfQTSHBDm3aH9u4C4WkZG2p9oyCXQU3g?usp=sharing) and [here](https://stackoverflow.com/questions/10530315/str-performance-in-python).

**Comment by GaelVaroquaux** [comment]
>     Also, what was the motivation for the directory name "cimg"?

> cimg is the repository for the new docker images of CircleCI:

Yes, I understood this later.

Sorry, my comment was a brain fart.

File: dirty_cat/minhash_encoder.py (hunk 8 above)
Comment (GaelVaroquaux) [reply dirty_cat/minhash_encoder.py:133]: I'd rather make sure that it renders well on the HTML rendering (which it might not if it is not valid numpy doc standard)

File: dirty_cat/test/test_gap_encoder.py [hunk 10]
```diff
@@ -159,36 +165,35 @@ def test_score(n_samples=70):
...
+                           match=r"handle_missing should be either "
+                                 r"'error' or 'zero_impute', got 'aaa'"):
             enc.fit_transform(observations)
 
 
-def profile_encoder(Encoder, init):
-    # not an unit test
-
-    from dirty_cat.datasets import fetch_employee_salaries
+def profile_encoder(init: str) -> None:
```
Comment (GaelVaroquaux) [inline dirty_cat/test/test_gap_encoder.py:190]: I wonder if we should not remove this code: if I understand well, it is dead code that is never run.

Let's consider this in a separate issue / PR, as I want to merge this

File: doc/_templates/class.rst [hunk 11]
```diff
@@ -9,7 +9,6 @@
 
 .. autoclass:: {{ objname }}
    :inherited-members:
-   :undoc-members:
```
Comment (GaelVaroquaux) [inline doc/_templates/class.rst:12]: I don't think that this should be removed. If someone contributes a class and forgets to add a docstring to a method (which could happen if we are not rigorous enough with our reviews) the resulting method will no longer be documented

**Review by GaelVaroquaux — CHANGES_REQUESTED** [review]
I'd like undoc-members to be added back, and then this is good to merge

**Comment by GaelVaroquaux** [comment]
After discussion: we will keep undoc-members removed because the current codebase is clean, find a way of hiding the generated attributes using numpydoc rather than napoleon (#309) and clever CSS.

File: dirty_cat/test/test_gap_encoder.py (hunk 10 above)
Comment (LilianBoulard) [reply dirty_cat/test/test_gap_encoder.py:190]: Added to the scope of #298

**Comment by GaelVaroquaux** [comment]
Wooot!
